tests/frontier/opcodes/test_calldatacopy.py

```diff
@@ -0,0 +1,176 @@
+                    dest_offset=0,
+                    offset=0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFA,
+                    size=0xFF,
+                )
+                + Op.SSTORE(key=0x0, value=Op.MLOAD(offset=0))
+                + Op.RETURN(offset=0, size=Op.MSIZE)
+            ),
+            b"\x04",
+            Account(storage={0x00: 0x00}),
+            Account(storage={0x00: 0x00}),
+        ),
+        (
+            (
+                Op.CALLDATACOPY(
+                    dest_offset=0,
+                    offset=0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFA,
+                    size=0x9,
+                )
+                + Op.SSTORE(key=0x0, value=Op.MLOAD(offset=0))
+                + Op.RETURN(offset=0, size=Op.MSIZE)
+            ),
+            b"\x05",
+            Account(storage={0x00: 0x00}),
+            Account(storage={0x00: 0x00}),
+        ),
+        (
+            (Op.SSTORE(key=0x1, value=0x1) + Op.PUSH1[0x1] + Op.PUSH1[0x2] + Op.CALLDATACOPY),
+            b"\x10",
+            Account(storage={0x01: 0x00}),
+            None,
+        ),
+        (
+            (
+                Op.JUMP(pc=0x5)
+                + Op.JUMPDEST
+                + Op.STOP
+                + Op.JUMPDEST
+                + Op.MSTORE8(offset=0x1F, value=0x42)
+                + Op.CALLDATACOPY(dest_offset=0x1F, offset=0x0, size=0x103)
+                + Op.MLOAD(offset=0x0)
+                + Op.DUP1
+                + Op.PUSH1[0x60]
+                + Op.JUMPI(pc=0x3, condition=Op.EQ)
+                + Op.SSTORE(key=0xFF, value=0xBADC0FFEE)
+            ),
+            b"\x11",
+            Account(storage={0xFF: 0xBADC0FFEE}),
+            None,
+        ),
+    ],
+    ids=[
+        "cdc 0 1 2",
+        "cdc 0 1 1",
+        "cdc 0 1 0",
+        "cdc 0 0 0",
+        "cdc 0 neg6 ff",
+        "cdc 0 neg6 9",
+        "underflow",
+        "sec",
+    ],
+)
+def test_calldatacopy(
+    state_test: StateTestFiller,
+    code: Bytecode,
+    fork: Fork,
+    tx_data: bytes,
+    pre: Alloc,
+    code_address_storage: Account,
+    to_address_storage: Account | None,
+):
+    """
+    Test `CALLDATACOPY` opcode.
+
+    Based on https://github.com/ethereum/tests/blob/ae4791077e8fcf716136e70fe8392f1a1f1495fb/src/GeneralStateTestsFiller/VMTests/vmTests/calldatacopyFiller.ym
+    """
+    code_address = pre.deploy_contract(code)
+    to = pre.deploy_contract(
+        code=(
+            Op.MSTORE(offset=0x0, value=0x1234567890ABCDEF01234567890ABCDEF0)
+            + Op.CALL(
+                0xFFFFFF, Op.ADD(0x1000, Op.CALLDATALOAD(offset=0x4)), 0x0, 0xF, 0x10, 0x20, 0x40
+            )
+            + Op.POP
+            + Op.SSTORE(key=0x0, value=Op.MLOAD(offset=0x20))
+            + Op.SSTORE(key=0x1, value=Op.MLOAD(offset=0x40))
+            + Op.STOP
+        ),
+        nonce=0,
+    )
+
+    tx = Transaction(
+        data=tx_data,
+        gas_limit=0x100_000_000,
```

Comment:
```suggestion
        gas_limit=100_000,
```
Yes should definitely be reduced, but I'm having a hard time figuring out why is it running out of gas.

It only happens in Frontier and Homestead, and it seems to be call on the root contract.

If you run with traces using `uv run fill --fork=Frontier "tests/frontier/opcodes/test_calldatacopy.py::test_calldatacopy[fork_Frontier-blockchain_test-cdc 0 1 2]" --evm-bin evm --traces`, the last call is trying to use 16 million gas for some reason:

```
Step 10:
{ 'depth': 1,
  'error': 'out of gas',
  'gas': '0x2bb13',
  'gasCost': '0x100002d',
  'memSize': 32,
  'op': 241,
  'opName': 'CALL',
  'pc': 38,
  'refund': 0,
  'stack': ['0x40', '0x20', '0x10', '0xf', '0x0', '0x1000', '0xffffff']}

Step 11:
{'error': 'out of gas', 'gasUsed': '0x2bb34', 'output': ''}
```

Reply:
Hey @pacrob I think we could still commit this suggestion, it should be possible now with the fix on line 152 👍